### PR TITLE
Add "MNEMONIC_REQUIRED" event for native wallet

### PR DESCRIPTION
**Pull request by majorhayes**
For the native wallet to work, it needs to be given the seed phrase. Similar to how KeepKey raises a PIN_REQUEST or PASSPHRASE_REQUEST event, this adds a new MNEMONIC_REQUIRED event that can be used to notify callers that the wallet is not initialized with a seed phrase.

* hdwallet-native needs to be able to be initialized without a mnemonic
* Raise MNEMONIC_REQUIRED whenever interacting with the seed
* Raise READY event after `loadDevice` is called to set the mnemonic

### CODE REVIEW

File: packages/hdwallet-native/src/native.ts
Comment (elmutt) [inline packages/hdwallet-native/src/native.ts:93]: How would this happen?  Can you give an example?

File: examples/sandbox/index.ts
Comment (kaladinlight) [inline examples/sandbox/index.ts:62]: Probably want to remove this debugging?

File: examples/sandbox/index.ts
Comment (kaladinlight) [inline examples/sandbox/index.ts:178]: Leave it auto initialize now that the event has been tested

File: package.json
Comment (kaladinlight) [inline package.json:28]: This was working as intended using tsconfig paths/references
```suggestion
    "build": "yarn tsc --build",
```

File: package.json
Comment (kaladinlight) [inline package.json:25]: and really this should probably look like
```suggestion
    "clean": "lerna run clean --scope @shapeshiftoss/* && rm -rf coverage test-report node_modules && yarn cache clean",
```

**Review by elmutt — APPROVED** [review]

File: packages/hdwallet-native/src/binance.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/binance.ts:41]: Would like to see this in core/binance.ts

File: packages/hdwallet-native/src/cosmos.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/cosmos.ts:40]: And this into `core/cosmos.ts`

File: packages/hdwallet-native/src/cosmos.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/cosmos.ts:81]: CosmosWallet needs the `needsMnemonic` wrapper where applicable

File: packages/hdwallet-native/src/ethereum.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/ethereum.ts:55]:
```suggestion
      return this.needsMnemonicAsync(!!this.#ethWallet, async () => this.#ethWallet.getAddress());
```

File: packages/hdwallet-native/src/adapter.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/adapter.ts:25]: If we generic to extend `HDWallet` we should probably update the rest of the wallets as well? A side question would be is this necessary?

File: packages/hdwallet-native/src/native.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/native.ts:48]: You mentioned the throw being problematic, should we just emit the event?

File: packages/hdwallet-native/src/native.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/native.ts:53]: And then here I feel like we should maybe not wrap in a try to allow the callback function to throw as it would be intended to instead of potentially eating the error.

File: packages/hdwallet-native/src/native.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/native.ts:33]: If we are going to doc comment the params, can we type it and provide a useful description if applicable?
`@param {boolean} hasMnemonic - useful description`

File: packages/hdwallet-native/src/native.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/native.ts:51]: And might as well document this one as well

File: packages/hdwallet-native/src/native.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/native.ts:20]: does this need to be a private field? not sure if it matters, but just curious as to your thought making it private.

**Comment by kaladinlight** [comment]
Also remember to bump version and update changelog before merging

File: examples/sandbox/index.ts
Comment (majorhayes) [reply examples/sandbox/index.ts:62]: Well, it's for the sandbox and I found it really helpful to show when an event was happening. I suppose I could have it just update some text on the page somewhere.

File: examples/sandbox/index.ts
Comment (majorhayes) [reply examples/sandbox/index.ts:178]: I took this out when the initialization was throwing an error for not having a mnemonic. Since I took out the error, you're right that we can probably have it auto initialize now.

File: package.json
Comment (majorhayes) [reply package.json:28]: It was definitely NOT working for me. Nothing was building at all, that's why I changed it. I wouldn't have noticed if it was working.

File: package.json
Comment (majorhayes) [reply package.json:25]: I don't think I changed that line but I can add it.

File: packages/hdwallet-native/src/binance.ts
Comment (majorhayes) [reply packages/hdwallet-native/src/binance.ts:41]: it's currently only implemented in native and keepkey so I wasn't sure if it was appropriate to have it in core if it doesn't apply to all wallets. What do you think?

File: packages/hdwallet-native/src/cosmos.ts
Comment (majorhayes) [reply packages/hdwallet-native/src/cosmos.ts:81]: Can't believe I missed that. Thanks.

File: packages/hdwallet-native/src/ethereum.ts
Comment (majorhayes) [reply packages/hdwallet-native/src/ethereum.ts:55]: The anonymous function doesn't need to be marked `async` because `getAddress` is async. That's why TypeScript isn't giving an error here because the return type is Promise.

File: packages/hdwallet-native/src/adapter.ts
Comment (majorhayes) [reply packages/hdwallet-native/src/adapter.ts:25]: The reason this is here is because the `events` property is unique to NativeHDWallet. If I leave it as a generic HDWallet, there's a compilation error for accessing a property that doesn't exist. `keyring.get` does have a type constraint that all generic types must extend HDWallet.

File: examples/sandbox/index.ts
Comment (kaladinlight) [reply examples/sandbox/index.ts:62]: Yeah I would rather it not create an alert box on any event personally. Do you know if the `debug.default` logs out the events to console as is already?

File: packages/hdwallet-native/src/native.ts
Comment (majorhayes) [reply packages/hdwallet-native/src/native.ts:48]: there's only one place that uses the synchronous version of the function and I didn't notice any problems with that, but I didn't do a ton of testing. I'll take a look at this again.

File: package.json
Comment (kaladinlight) [reply package.json:28]: What were you seeing then? It is using the references to build packages in order based on any dependencies.

File: package.json
Comment (kaladinlight) [reply package.json:25]: I just noticed it when looking at the build script. Thanks

File: packages/hdwallet-native/src/native.ts
Comment (majorhayes) [reply packages/hdwallet-native/src/native.ts:53]: I'll take a look at this again. So far in my tests, any of these functions throwing an error was not being handled by axiom and was causing a crash, but I see your point about separating the error from `needsMnemonic` from the callback itself.

File: packages/hdwallet-native/src/native.ts
Comment (majorhayes) [reply packages/hdwallet-native/src/native.ts:33]: I didn't meant to leave that in.

File: packages/hdwallet-native/src/binance.ts
Comment (kaladinlight) [reply packages/hdwallet-native/src/binance.ts:41]: That was the pattern I moved towards for native and think it makes more sense since it should be the same for all wallets. Also if we add to ledger or trezor, we won't need to copy it there as well. Would be my preference.

File: packages/hdwallet-native/src/native.ts
Comment (majorhayes) [reply packages/hdwallet-native/src/native.ts:20]: I wanted it to be read-only at runtime. Probably not a big deal either way.

**Comment by majorhayes** [comment]
> Also remember to bump version and update changelog before merging

About that... I'm not sure that it's the right thing to do to do a version bump in a feature PR. I think it's better to have a separate version bump specifically when we're ready for a release to go out. It's not uncommon for two PRs to be open at the time and if both have version bumps you have to resolve conflicts depending on which gets merged first.

I would think that it would be better to just merge the PR and then do a release PR to bump the version.

We have the same issue with axiom packages.

File: packages/hdwallet-native/src/ethereum.ts
Comment (kaladinlight) [reply packages/hdwallet-native/src/ethereum.ts:55]: Yeah that makes sense. It "looks" more right in ref to `needsMnemonicAsync` without having to know `getAddress` return type. I guess like you said typescript would complain if need be. Nitpick for looks overall. Your call.

**Comment by kaladinlight** [comment]
> > Also remember to bump version and update changelog before merging
> 
> About that... I'm not sure that it's the right thing to do to do a version bump in a feature PR. I think it's better to have a separate version bump specifically when we're ready for a release to go out. It's not uncommon for two PRs to be open at the time and if both have version bumps you have to resolve conflicts depending on which gets merged first.
> 
> I would think that it would be better to just merge the PR and then do a release PR to bump the version.
> 
> We have the same issue with axiom packages.

I can get behind that for sure. Just want to make sure we don't forget it as I have done that before haha

File: packages/hdwallet-native/src/native.ts
Comment (kaladinlight) [reply packages/hdwallet-native/src/native.ts:53]: Yeah, we should probably keep track of where axiom doesn't handle hdwallet errors correctly and actually address them

File: packages/hdwallet-native/src/native.ts
Comment (kaladinlight) [reply packages/hdwallet-native/src/native.ts:20]: ok, just curious more than anything

File: examples/sandbox/index.ts
Comment (majorhayes) [reply examples/sandbox/index.ts:62]: I went ahead and put some time into implementing a mnemonic_required modal in sandbox and added an "event log" section rather than relying on "alert".

File: examples/sandbox/index.ts
Comment (majorhayes) [reply examples/sandbox/index.ts:178]: I put it back in, but I had to remove a `wallet.initialize()` line in the "on("change")` handler for the select at the top to avoid always popping up the mnemonic required modal.

File: package.json
Comment (kaladinlight) [inline package.json:33]: so this actually used to `dev:sandbox` I had accidentally removed it a little while back and it is still referenced in the readme. can we use this instead?
```suggestion
    "dev:sandbox": "lerna run dev --scope hdwallet-sandbox --parallel --include-filtered-dependencies",
```

File: packages/hdwallet-native/src/native.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/native.ts:128]: Don't need this anymore

File: packages/hdwallet-native/src/native.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/native.ts:98]: Can this stay as super to show we are calling a function from an extended class.

File: packages/hdwallet-native/src/adapter.ts
Comment (kaladinlight) [inline packages/hdwallet-native/src/adapter.ts:33]: We don't really need to wrap this is the mnemonic check now with the addition of `needsMnemonic` in `wallet.initialize`. Thoughts?

File: packages/hdwallet-native/src/native.ts
Comment (majorhayes) [reply packages/hdwallet-native/src/native.ts:128]: my IDE automatically added it I guess because it's technically part of the interface, but I'll remove it.

File: packages/hdwallet-native/src/native.ts
Comment (majorhayes) [reply packages/hdwallet-native/src/native.ts:98]: I read up on this before the change, and unless overridden, they are basically the same thing. Since it's the child, `this` works just fine. It's not really worth arguing about though.

File: packages/hdwallet-native/src/adapter.ts
Comment (majorhayes) [reply packages/hdwallet-native/src/adapter.ts:33]: that's actually WHY we need it, because otherwise it'll immediate raise the mnemonic required event immediately on initializing the adapter, which is what I didn't want to happen.

Comment (majorhayes) [reply packages/hdwallet-native/src/adapter.ts:33]: BTW, in my head I'm considering "adapter.initialize()" as a separate function from "wallet.initialize". I see the former as setting up the adapter and the latter as setting up the wallet. In our case, setting up the adapter itself shouldn't require the mnemonic, only initializing the wallet itself should. Maybe I'm thinking about that wrong, or not how it was intended?

Comment (kaladinlight) [reply packages/hdwallet-native/src/adapter.ts:33]: No, you are good, I confused myself. You are correct

File: packages/hdwallet-native/src/native.ts
Comment (kaladinlight) [reply packages/hdwallet-native/src/native.ts:98]: Sure, I would just say we are using super to call all parent class function except for this function now is all. I will leave it at that.

**Review by kaladinlight — APPROVED** [review]

**Pull request merged**
